### PR TITLE
parse more dot operators, parse .= with assignment precedence

### PR DESCRIPTION
**Pull request by stevengj**
For 0.6, the plan in #16285 is to have `a .+ b` etcetera automatically correspond to `broadcast(+, a, b)`, and be fused with other "dot calls".   Also, we are planning to have `x .= ...` etcetera turn into `broadcast!` for in-place fused loops.

As a first step towards that goal, this PR:
- Changes the precedence of `.=` to be assignment-like, rather than comparison-like.
- Makes (nearly) every operator "dottable".  The exceptions are operators with very special syntactic meanings, like `$` and `||`.   (Some of these could be allowed, but I wanted to be conservative to start with).

I'd ideally like to get this merged in 0.5.   The reason is that, once 0.6 rolls around, we will want a `@compat` macro to make some of #16285's vectorized syntax work (albeit suboptimally) in 0.5, and it will be hard to do that unless the operators parse.

### CODE REVIEW

**Comment by stevengj** [comment]
(I don't think we should document the new operators, yet, because we don't really want people to start assigning methods to them in 0.5 ... since in 0.6 the `broadcast` methods will be defined automatically given a method for the non-dotted operator.)

**Comment by stevengj** [comment]
As far as I can tell, no existing packages use `.=`, so the change in precedence should be fairly safe.   In 0.6, `.=` will be a (broadcasting) assignment operator so it will not be possible to add methods to it anyway.

**Comment by StefanKarpinski** [comment]
:+1:

**Comment by JeffBezanson** [comment]
It seems like we could allow `.&` and `.|`.

**Comment by stevengj** [comment]
Added `.&` and `.|`

**Comment by stevengj** [comment]
@JeffBezanson, I'm bugged by `.!`.   As I understand it, in order to parse `.!=` as an operator the parser requires the prefix `.!` to also be parsed as an operator, which is why `.!` has to be an operator.

However, the `.!` operator was parsed with `prec-comparison`, i.e. as a *binary* operator with comparison precedence.  I kept this behavior, but it seems weird.  Shouldn't we parse `.!` as a unary operator if we parse it at all?  Or would that screw up parsing of `.!=`?

**Comment by JeffBezanson** [comment]
We could handle that by rejecting `.!` at the end of `read-operator`. That is already done for `--`.

**Comment by stevengj** [comment]
Or maybe we should just allow unary operators to be dotted?  So that `.!x` would turn (in 0.6) into `broadcast((!), x)`?

**Comment by stevengj** [comment]
(It seems like `--` should nowadays be a `prec-plus` operator, similar to `++`.)

**Comment by stevengj** [comment]
Okay, I took Jeff's suggestion and reject `.!` in the parser, like `--`.   Just adding dots before unary operators didn't work (seems to require some deeper changes in the parser), so I punted on that for now.   With unary operators the need for a dot prefix is much less, because you can always write `(!).(array)`.

**Comment by stevengj** [comment]
Seems like the libgit2 Travis test failure must be unrelated?

**Comment by stevengj** [comment]
Restarting checks since the Travis failure on libgit2 should hopefully have been fixed by 74cf1b12d75e195a38c65f1a6ac6f75b0203dd46.

**Comment by stevengj** [comment]
Rebased and added NEWS.

**Comment by tkelman** [comment]
Should this have any parsing tests?

**Comment by andyferris** [comment]
I think having `.-`, `.!`, etc as unary operators would make a lot of sense, but I understand you said it might be hard to implement.

I can currently do `-v` and `(-).(v)`, so it seems odd I that *can't* do `.-v`.

**Comment by stevengj** [comment]
@tkelman, added some tests.

**Comment by stevengj** [comment]
@andyferris, I don't think it should be *hard* to implement more dotted unary operators, it is just that we are so close to the 0.5 release that I don't want to dig deep into the parser right now.

**Comment by stevengj** [comment]
Okay to merge?

**Comment by tkelman** [comment]
lgtm but I'm the wrong person to review the scheme code

**Comment by stevengj** [comment]
@JeffBezanson, looks like it's up to you to hit the "merge" button?

**Comment by StefanKarpinski** [comment]
@JeffBezanson is on vacation, so I'd recommend merging if you're happy with it and we can do a post-merge review when he gets back.

**Comment by vtjnash** [comment]
lgtm. do these deparse (print) correctly?

(also, there's now a trivial merge conflict in test/core, so needs a rebase)

File: test/core.jl [hunk 1]
```diff
@@ -4462,3 +4462,11 @@ end
 f17147(::Tuple) = 1
 f17147{N}(::Vararg{Tuple,N}) = 2
 @test f17147((), ()) == 2
+
+# PR #17393
```
Comment (vtjnash) [inline test/core.jl]: put this in `test/parse`?

Comment (stevengj) [inline test/core.jl]: ok

**Comment by stevengj** [comment]
Getting an unrelated segfault?
```
signal (11): Segmentation fault
while loading /tmp/julia/share/julia/test/subarray.jl, in expression starting on line 280
processFDE<register_eh_frames(uint8_t*, size_t)::<lambda(char const*)> > at /home/travis/build/JuliaLang/julia/src/debuginfo.cpp:1363 [inlined]
processFDEs<register_eh_frames(uint8_t*, size_t)::<lambda(char const*)> > at /home/travis/build/JuliaLang/julia/src/debuginfo.cpp:1380 [inlined]
register_eh_frames at /home/travis/build/JuliaLang/julia/src/debuginfo.cpp:1628
```
@yuyichao, could this be related to #16795

**Comment by yuyichao** [comment]
Unlikely #16795 but certainly related to unwind info parsing.

**Comment by stevengj** [comment]
@vtjnash, yes, the new dotted binary operators print fine.  The assignment syntax `x .= y` prints as an `Expr` object, but since `x .+= y` in Julia 0.4 also prints as an `Expr` object I think that is fine... we should fix printing of dotted assignments once this actually does something.

**Comment by mbauman** [comment]
Oooh, this is cool.  Any thoughts here on decoupling the precedence of `.|` and `.&` from `|` and `&` to support a use case like #5187?  That'd be something that'd be good to do in the same cycle that they're introduced since precedences are very difficult to change.

**Comment by stevengj** [comment]
@mbauman, I find #5187 very hard to follow because of the length of the thread.  I would suggest making a PR that changes the precedence, and summarize in the PR what you think are the main arguments in favor from #5187.

**Comment by tkelman** [comment]
one note out of this is that we could deprecate elementwise `|` now that `.|` is a general version.

**Comment by KristofferC** [comment]
Did this PR break the following?

```jl
julia> Base.|>(a, b) = a+b
ERROR: syntax: ">(a,b)" is not a valid function argument name
 in eval(::Module, ::Any) at ./boot.jl:234
 in macro expansion at ./REPL.jl:92 [inlined]
 in (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:46
```

(found from Gtk.jl stopped parsing):

**Comment by stevengj** [comment]
The dot operators are now parsed, but the don't do anything different from 0.4 in this PR. 

Deprications will have to wait until 0.6 when dot operators are sugar for a broadcast.

**Comment by KristofferC** [comment]
Ok, I guess something else made that syntax fail to parse because it worked in 0.4 and it is currently failing. That fact that the error message says that `>(a,b)` is not a valid function name (the `.|` part being removed from the error message) made me suspect that this PR had something to do with it,

**Comment by stevengj** [comment]
@KristofferC, sorry, I was replying to @tkelman.   Probably this PR broke `Base.|>`.  You should now write `Base.:|>`, the same way that you write `Base.:+`

**Comment by tkelman** [comment]
Also broke IntervalConstraintProgramming and LibExpat
